File: android/app/src/main/kotlin/com/orialthq/ori_beauty/share/IncomingShareStore.kt
package com.orialthq.ori_beauty.share

import android.content.Context
import java.io.File
import org.json.JSONArray
import org.json.JSONObject

class IncomingShareStore(context: Context) {
    private val attachmentDirectory =
        File(context.noBackupFilesDir, IncomingShareIngestor.ATTACHMENT_DIRECTORY_NAME)
    private val preferences =
        context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)

    @Synchronized
    fun append(payload: IncomingSharePayload): Boolean {
        val pending = readArray()
        val item =
            JSONObject()
                .put("id", payload.id)
                .put("receivedAtEpochMs", payload.receivedAtEpochMs)
                .put("sharedText", payload.sharedText)
                .put("discoveredUrl", payload.discoveredUrl)
                .put("sourcePackage", payload.sourcePackage)
                .put("mimeType", payload.mimeType)
                .put("wasTruncated", payload.wasTruncated)
                .put("originalLength", payload.originalLength)
                .put("shareKind", payload.shareKind)
                .put(
                    "attachments",
                    JSONArray().apply {
                        payload.attachments.forEach { attachment ->
                            put(
                                JSONObject()
                                    .put("id", attachment.id)
                                    .put("filePath", attachment.filePath)
                                    .put("mimeType", attachment.mimeType)
                                    .put("byteSize", attachment.byteSize)
                                    .put("width", attachment.width)
                                    .put("height", attachment.height)
                                    .put("sha256", attachment.sha256),
                            )
                        }
                    },
                )

        pending.put(item)
        return writeArray(pending)
    }

    @Synchronized
    fun pending(): List<Map<String, Any?>> {
        val pending = readArray()
        return buildList {
            for (index in 0 until pending.length()) {
                val item = pending.getJSONObject(index)
                val attachments =
                    item.optJSONArray("attachments")?.toAttachmentMaps().orEmpty()
                add(
                    mapOf(
                        "id" to item.getString("id"),
                        "receivedAtEpochMs" to item.getLong("receivedAtEpochMs"),
                        "sharedText" to item.getString("sharedText"),
                        "discoveredUrl" to item.optString("discoveredUrl").ifBlank { null },
                        "sourcePackage" to
                            item.optString("sourcePackage").ifBlank { null },
                        "mimeType" to item.optString("mimeType", "text/plain"),
                        "wasTruncated" to item.optBoolean("wasTruncated", false),
                        "originalLength" to
                            item.optInt(
                                "originalLength",
                                item.getString("sharedText").length,
                            ),
                        "shareKind" to
                            item.optString("shareKind").ifBlank {
                                if (attachments.isEmpty()) SHARE_KIND_TEXT else SHARE_KIND_IMAGE
                            },
                        "attachments" to attachments,
                    ),
                )
            }
        }
    }

    @Synchronized
    fun loadAppSnapshot(): String? = preferences.getString(KEY_APP_SNAPSHOT, null)

    @Synchronized
    fun saveAppSnapshot(snapshot: String): Boolean {
        if (snapshot.isBlank()) {
            return false
        }
        return preferences
            .edit()
            .putString(KEY_APP_SNAPSHOT, snapshot)
            .commit()
    }

    @Synchronized
    fun acknowledge(ids: Collection<String>): Boolean {
        if (ids.isEmpty()) {
            return true
        }
        val pending = readArray()
        val remaining = JSONArray()
        val acknowledgedAttachmentPaths = mutableListOf<String>()
        for (index in 0 until pending.length()) {
            val item = pending.getJSONObject(index)
            if (item.getString("id") !in ids) {
                remaining.put(item)
            } else {
                val attachments = item.optJSONArray("attachments")
                if (attachments != null) {
                    for (attachmentIndex in 0 until attachments.length()) {
                        attachments
                            .optJSONObject(attachmentIndex)
                            ?.optString("filePath")
                            ?.takeIf(String::isNotBlank)
                            ?.let(acknowledgedAttachmentPaths::add)
                    }
                }
            }
        }
        if (!writeArray(remaining)) {
            return false
        }
        acknowledgedAttachmentPaths.forEach(::deleteManagedAttachment)
        return true
    }

    private fun readArray(): JSONArray {
        val raw = preferences.getString(KEY_PENDING_SHARES, null)
        return if (raw.isNullOrBlank()) JSONArray() else runCatching { JSONArray(raw) }
            .getOrDefault(JSONArray())
    }

    private fun writeArray(value: JSONArray): Boolean {
        return preferences
            .edit()
            .putString(KEY_PENDING_SHARES, value.toString())
            .commit()
    }

    private fun JSONArray.toAttachmentMaps(): List<Map<String, Any?>> {
        return buildList {
            for (index in 0 until length()) {
                val attachment = optJSONObject(index) ?: continue
                add(
                    mapOf(
                        "id" to attachment.optString("id"),
                        "filePath" to attachment.optString("filePath"),
                        "mimeType" to attachment.optString("mimeType"),
                        "byteSize" to attachment.optLong("byteSize"),
                        "width" to attachment.optInt("width"),
                        "height" to attachment.optInt("height"),
                        "sha256" to attachment.optString("sha256"),
                    ),
                )
            }
        }
    }

    private fun deleteManagedAttachment(filePath: String) {
        val root = runCatching { attachmentDirectory.canonicalFile }.getOrNull() ?: return
        val file = runCatching { File(filePath).canonicalFile }.getOrNull() ?: return
        if (file.parentFile == root) {
            file.delete()
        }
    }

    companion object {
        private const val PREFERENCES_NAME = "incoming_share_store"
        private const val KEY_PENDING_SHARES = "pending_shares_v1"
        private const val KEY_APP_SNAPSHOT = "app_snapshot_v1"
    }
}

File: ios/Runner/Share/IncomingShareStore.swift
import Foundation

/// Mirror of `share/IncomingShareStore.kt` for the pending-share queue.
///
/// Android keeps the queue in `SharedPreferences`; iOS keeps it in a JSON file next
/// to the attachments. The app snapshot is *not* handled here — iOS already stores it
/// through `AppSnapshotFileStore`, so this type covers only the queue that
/// `drainPendingShares` and `acknowledgeShares` operate on.
final class IncomingShareStore {
  static let shared = IncomingShareStore()

  private let fileManager = FileManager.default
  private let lock = NSLock()

  private lazy var queueURL: URL? = {
    guard let base = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
    else {
      return nil
    }
    do {
      try fileManager.createDirectory(
        at: base,
        withIntermediateDirectories: true,
        attributes: nil
      )
      return base.appendingPathComponent("incoming_share_pending_v1.json", isDirectory: false)
    } catch {
      return nil
    }
  }()

  private lazy var attachmentDirectory: URL? = {
    guard let base = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
    else {
      return nil
    }
    return base.appendingPathComponent(
      IncomingShareIngestor.attachmentDirectoryName,
      isDirectory: true
    )
  }()

  private init() {}

  @discardableResult
  func append(_ payload: IncomingSharePayload) -> Bool {
    lock.lock()
    defer { lock.unlock() }

    var pending = readArray()
    pending.append([
      "id": payload.id,
      "receivedAtEpochMs": payload.receivedAtEpochMs,
      "sharedText": payload.sharedText,
      "discoveredUrl": payload.discoveredUrl as Any,
      "sourcePackage": payload.sourcePackage as Any,
      "mimeType": payload.mimeType,
      "wasTruncated": payload.wasTruncated,
      "originalLength": payload.originalLength,
      "shareKind": payload.shareKind,
      "attachments": payload.attachments.map { attachment in
        [
          "id": attachment.id,
          "filePath": attachment.filePath,
          "mimeType": attachment.mimeType,
          "byteSize": attachment.byteSize,
          "width": attachment.width,
          "height": attachment.height,
          "sha256": attachment.sha256,
        ] as [String: Any]
      },
    ])
    return writeArray(pending)
  }

  /// Platform maps consumed by `IncomingShare.fromPlatformMap` on the Dart side.
  func pending() -> [[String: Any]] {
    lock.lock()
    defer { lock.unlock() }

    return readArray().compactMap { item in
      guard let id = item["id"] as? String, !id.isEmpty,
        let receivedAtEpochMs = item["receivedAtEpochMs"] as? Int64 ?? (item["receivedAtEpochMs"] as? Int).map(Int64.init),
        let sharedText = item["sharedText"] as? String
      else {
        return nil
      }
      let attachments = attachmentMaps(item["attachments"])
      return [
        "id": id,
        "receivedAtEpochMs": receivedAtEpochMs,
        "sharedText": sharedText,
        "discoveredUrl": (item["discoveredUrl"] as? String)?.nilWhenBlank as Any,
        "sourcePackage": (item["sourcePackage"] as? String)?.nilWhenBlank as Any,
        "mimeType": (item["mimeType"] as? String)?.nilWhenBlank ?? "text/plain",
        "wasTruncated": item["wasTruncated"] as? Bool ?? false,
        "originalLength": item["originalLength"] as? Int ?? sharedText.count,
        "shareKind": (item["shareKind"] as? String)?.nilWhenBlank
          ?? (attachments.isEmpty ? ShareKind.text : ShareKind.image),
        "attachments": attachments,
      ]
    }
  }

  @discardableResult
  func acknowledge(ids: [String]) -> Bool {
    guard !ids.isEmpty else { return true }

    lock.lock()
    defer { lock.unlock() }

    let acknowledged = Set(ids)
    var remaining: [[String: Any]] = []
    var acknowledgedAttachmentPaths: [String] = []

    for item in readArray() {
      guard let id = item["id"] as? String else { continue }
      if acknowledged.contains(id) {
        for attachment in attachmentMaps(item["attachments"]) {
          if let path = (attachment["filePath"] as? String)?.nilWhenBlank {
            acknowledgedAttachmentPaths.append(path)
          }
        }
      } else {
        remaining.append(item)
      }
    }

    guard writeArray(remaining) else {
      return false
    }
    acknowledgedAttachmentPaths.forEach(deleteManagedAttachment)
    return true
  }

  private func attachmentMaps(_ raw: Any?) -> [[String: Any]] {
    guard let items = raw as? [[String: Any]] else { return [] }
    return items.compactMap { attachment in
      guard let id = attachment["id"] as? String,
        let filePath = attachment["filePath"] as? String,
        let mimeType = attachment["mimeType"] as? String,
        let sha256 = attachment["sha256"] as? String
      else {
        return nil
      }
      let byteSize = attachment["byteSize"] as? Int64
        ?? (attachment["byteSize"] as? Int).map(Int64.init)
        ?? 0
      return [
        "id": id,
        "filePath": filePath,
        "mimeType": mimeType,
        "byteSize": byteSize,
        "width": attachment["width"] as? Int ?? 0,
        "height": attachment["height"] as? Int ?? 0,
        "sha256": sha256,
      ]
    }
  }

  private func readArray() -> [[String: Any]] {
    guard let queueURL,
      let data = try? Data(contentsOf: queueURL),
      !data.isEmpty,
      let decoded = try? JSONSerialization.jsonObject(with: data),
      let array = decoded as? [[String: Any]]
    else {
      return []
    }
    return array
  }

  private func writeArray(_ value: [[String: Any]]) -> Bool {
    guard let queueURL,
      let data = try? JSONSerialization.data(withJSONObject: value)
    else {
      return false
    }
    do {
      try data.write(to: queueURL, options: [.atomic, .completeFileProtection])
      var resourceValues = URLResourceValues()
      resourceValues.isExcludedFromBackup = true
      var mutableURL = queueURL
      try? mutableURL.setResourceValues(resourceValues)
      return true
    } catch {
      return false
    }
  }

  private func deleteManagedAttachment(filePath: String) {
    guard let attachmentDirectory else { return }
    let root = URL(fileURLWithPath: attachmentDirectory.path).standardizedFileURL
    let file = URL(fileURLWithPath: filePath).standardizedFileURL
    guard file.deletingLastPathComponent() == root else { return }
    try? fileManager.removeItem(at: file)
  }
}

extension String {
  fileprivate var nilWhenBlank: String? {
    let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed.isEmpty ? nil : self
  }
}
